OpenDreamShared/Resources/DMIParser.cs

```diff
@@ -54,6 +57,81 @@ public class ParsedDMIFrame {
             public float Delay;
         }
 
+        private static uint _readBigEndianUint32(BinaryReader reader)
+        {
+            byte[] bytes = reader.ReadBytes(4);
+            Array.Reverse(bytes); //Little to Big-Endian
+            return BitConverter.ToUInt32(bytes);
+        }
+
+        private static (uint, string) _readChunk(BinaryReader reader)
+        {
+            uint chunkLength = _readBigEndianUint32(reader);
+            string chunkType = new string(reader.ReadChars(4));
+
+            return (chunkLength, chunkType);
+        }
+
+        private static (uint, uint) _readIHDRChunk(BinaryReader reader, MemoryStream stream)
+        {
+            var (chunkLength, chunkType) = _readChunk(reader);
+            if (chunkType != "IHDR")
+            {
+                throw new ArgumentException($"Expected IHDR chunk, got ${chunkType} instead.");
+            }
+
+            var width = _readBigEndianUint32(reader);
+            var height = _readBigEndianUint32(reader);
+
+            stream.Seek(chunkLength + 4 - sizeof(uint) * 2, SeekOrigin.Current);
+
+            return (width, height);
+        }
+
+        public static (string, uint, uint) ReadDMIDescription(byte[] bytes) {
+            MemoryStream stream = new MemoryStream(bytes);
+            BinaryReader reader = new BinaryReader(stream);
+
+            stream.Seek(8, SeekOrigin.Begin);
+
+            var (width, height) = _readIHDRChunk(reader, stream);
```

Comment:
I don't like the IHDR reader being used in the zTXt reader, could you split them up?

Reply:
I'm not sure what you mean. It's necessary to somehow obtain the width from the IHDR chunk because `ParseDMIDescription` needs it.

Reply:
`ReadDMIDescription` is the zTXt reader, and shouldn't be returning data from other chunks.

With this being moved to shared, I don't see any reason for code using DMIParser to have to use both ReadDMIDescription and ParseDMIDescription to parse a DMI. Maybe instead there could be a `ParseDMI(byte[] data)` method that reads the chunks, and returns a ParsedDMIDescription?

Reply:
`ReadDMIDescription` is back to only dealing with zTXt. `icon_states` now uses the new `GetIconStatesFromDescription`, which exclusively cares about icon states. `ParseDMIDescription` is back to only being used by the client. Is this okay?

Reply:
Yes, that's good